operators/pkg/tenant-controller/tenant_controller.go

```diff
@@ -307,6 +311,86 @@ func (r *TenantReconciler) checkValidWorkspaces(ctx context.Context, tn *crownla
 	return tenantExistingWorkspaces, workspaces, err
 }
 
+// deleteClusterNamespace deletes the namespace for the tenant, if it fails then it returns an error.
+func (r *TenantReconciler) deleteClusterNamespace(ctx context.Context, tn *crownlabsv1alpha2.Tenant, nsName string) (err error) {
+	ns := v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: nsName}}
+
+	nsErr := utils.EnforceObjectAbsence(ctx, r.Client, &ns, "personal namespace")
+
+	if nsErr != nil {
+		klog.Errorf("Error when deleting namespace of tenant %s -> %s", tn.Name, nsErr)
+	}
+
+	return nsErr
+}
+
+// enforceNamespaceKeepAliveOrDelete deletes the namespace if it should be deleted.
+func (r *TenantReconciler) enforceNamespaceKeepAliveOrDelete(ctx context.Context, tn *crownlabsv1alpha2.Tenant, nsName string, tenantWorkspaceKeepAlive time.Duration) (keepNsOpen bool, err error) {
+	// We check to see if last login was more than tenantWorkspaceKeepAlive in the past:
+	// if so, temporarily delete the namespace. We assume that a lastLogin of 0 occurs when a user is first created
+
+	// Calculate time elapsed since lastLogin (now minus lastLogin in seconds)
+	sPassed := time.Since(tn.Spec.LastLogin.Time)
+
+	klog.Infof("Last login of tenant %s was %s ago", tn.Name, sPassed)
+
+	// Attempt to get instances in current namespace
+	list := &crownlabsv1alpha2.InstanceList{}
+
+	if err := r.List(context.Background(), list, client.InNamespace(nsName)); err != nil {
+		return true, err
+	}
+
+	if sPassed > tenantWorkspaceKeepAlive { // seconds
+		klog.Infof("Over %s elapsed since last login of tenant %s: attempting to delete tenant namespace if not already deleted", tenantWorkspaceKeepAlive, tn.Name)
+		if len(list.Items) == 0 {
+			klog.Infof("No instances in %s: workspace can be deleted", nsName)
+			keepNsOpen = false
```

Comment:
This variable is never initialized, hence it is always false. BTW, when possible prefer to return early rather than setting variables, to simplify the workflow and make your intentions clear.

Reply:
Resolved in [3b5ee71](https://github.com/netgroup-polito/CrownLabs/pull/821/commits/3b5ee717263c30db8a3b0cd9daa506d04396379e)